stacker/plan2.py

```diff
@@ -0,0 +1,345 @@
+        """Returns a "filtered" version of this graph."""
+        return Graph(steps=self.steps, dag=self.dag.filter(step_names))
+
+    def topological_sort(self):
+        nodes = self.dag.topological_sort()
+        return [self.steps[step_name] for step_name in nodes]
+
+
+class Plan(object):
+    """A collection of :class:`Step` objects to execute.
+    The :class:`Plan` helps organize the steps needed to execute a particular
+    action for a set of :class:`stacker.stack.Stack` objects. When initialized
+    with a set of steps, it will first build a Directed Acyclic Graph from the
+    steps and their dependencies.
+    Args:
+        description (str): description of the plan.
+        graph (:class:`Graph`): a graph of steps.
+    """
+
+    def __init__(self, description=None, graph=None):
+        self.id = uuid.uuid4()
+        self.description = description
+        self.graph = graph
+
+    def outline(self, level=logging.INFO, message=""):
+        """Print an outline of the actions the plan is going to take.
+        The outline will represent the rough ordering of the steps that will be
+        taken.
+        Args:
+            level (int, optional): a valid log level that should be used to log
+                the outline
+            message (str, optional): a message that will be logged to
+                the user after the outline has been logged.
+        """
+        steps = 1
+        logger.log(level, "Plan \"%s\":", self.description)
+        for step in self.steps:
+            logger.log(
+                level,
+                "  - step: %s: target: \"%s\", action: \"%s\"",
+                steps,
+                step.short_name,
+                step.fn.__name__,
+            )
+            steps += 1
+
+        if message:
+            logger.log(level, message)
+
+    def dump(self, directory, context, provider=None):
+        logger.info("Dumping \"%s\"...", self.description)
+        directory = os.path.expanduser(directory)
+        if not os.path.exists(directory):
+            os.makedirs(directory)
+
+        def walk_func(step):
+            step.stack.resolve(
+                context=context,
+                provider=provider,
+            )
+            blueprint = step.stack.blueprint
+            filename = stack_template_key_name(blueprint)
+            path = os.path.join(directory, filename)
+
+            blueprint_dir = os.path.dirname(path)
+            if not os.path.exists(blueprint_dir):
+                os.makedirs(blueprint_dir)
+
+            logger.info("Writing stack \"%s\" -> %s", step.name, path)
+            with open(path, "w") as f:
+                f.write(blueprint.rendered)
+
+            return True
+
+        return self.graph.walk(walk_func)
+
+    def execute(self, **kwargs):
+        return self.walk(**kwargs)
+
+    def walk(self):
+        """Walks each step in the underlying graph, in topological order."""
+
+        def walk_func(step):
```

Comment:
I think we definitely should make that change, since it's annoying when a stack throws an exception, and you have no context about what stack has the issue. But we should probably make that change separate from this PR, since that's not existing behavior.

We'll also need to think about how we handle exceptions in a parallelized world, since the current method of just throwing an exception and aborting won't work with multi-threads/processes.